tests/testthat/_snaps/tree.md

```diff
@@ -0,0 +1,219 @@
+      └─children: <list>
+        ├─S3<shiny.tag>
+        │ ├─name: "label"
+        │ ├─attribs: <list>
+        │ │ ├─class: "control-label"
+        │ │ ├─id: "test-label"
+        │ │ └─for: "test"
+        │ └─children: <list>
+        │   └─"Input Label"
+        └─S3<shiny.tag>
+          ├─name: "input"
+          ├─attribs: <list>
+          │ ├─class: "js-range-slider"
+          │ ├─id: "test"
+          │ ├─data-skin: "shiny"
+          │ ├─data-min: "0"
+          │ ├─data-max: "1"
+          │ ├─data-from: "0.5"
+          │ ├─data-step: "0.01"
+          │ ├─data-grid: "true"
+          │ ├─data-grid-num: 10
+          │ ├─data-grid-snap: "false"
+          │ ├─data-prettify-separator: ","
+          │ ├─data-prettify-enabled: "true"
+          │ ├─data-keyboard: "true"
+          │ └─data-data-type: "number"
+          └─children: <list>
+
+# Max depth and length can be enforced
+
+    Code
+      deep_list <- list(list(id = "b", val = 1, children = list(list(id = "b1", val = 2.5),
+      list(id = "b2", val = 8, children = list(list(id = "b21", val = 4))))), list(
+        id = "a", val = 2))
+      tree(deep_list, max_depth = 1)
+    Output
+      <list>
+      ├─<list>...
+      └─<list>...
+    Code
+      tree(deep_list, max_depth = 2)
+    Output
+      <list>
+      ├─<list>
+      │ ├─id: "b"
+      │ ├─val: 1
+      │ └─children: <list>...
+      └─<list>
+        ├─id: "a"
+        └─val: 2
+    Code
+      tree(deep_list, max_depth = 3)
+    Output
+      <list>
+      ├─<list>
+      │ ├─id: "b"
+      │ ├─val: 1
+      │ └─children: <list>
+      │   ├─<list>...
+      │   └─<list>...
+      └─<list>
+        ├─id: "a"
+        └─val: 2
+    Code
+      tree(deep_list, max_length = 0)
+    Output
+      ... 
+    Code
+      tree(deep_list, max_length = 2)
+    Output
+      <list>
+      ├─<list>
+      ... 
+    Code
+      tree(deep_list, max_depth = 1, max_length = 4)
+    Output
+      <list>
+      ├─<list>...
+      └─<list>...
+
+# Missing values are caught and printed properly
+
+    Code
+      tree(list(`null-element` = NULL, `NA-element` = NA))
+    Output
+      <list>
+      ├─null-element: <NULL>
+      └─NA-element: NA
+
+# non-named elements in named list
+
+    Code
+      tree(list(a = 1, "el w/o id"))
+    Output
+      <list>
+      ├─a: 1
+      └─"el w/o id"
+
+# Attributes are properly displayed as special children nodes
+
+    Code
+      list_w_attrs <- structure(list(structure(list(id = "a", val = 2), level = 2,
+      name = "first child"), structure(list(id = "b", val = 1, children = list(list(
+        id = "b1", val = 2.5))), level = 2, name = "second child", class = "custom-class"),
+      level = "1", name = "root"))
+      tree(list_w_attrs, show_attributes = TRUE)
+    Output
+      <list>
+      ├─<list>
+      │ ├─id: "a"
+      │ ├─val: 2
+      │ ├┄<attr>names: ["id", "val"]
+      │ ├┄<attr>level: 2
+      │ └┄<attr>name: "first child"
```

Comment:
What do you think of nesting attributes, like this?

```suggestion
      ├─<list>
      │ ├─id: "a"
      │ ├─val: 2
      │ ├┄attributes
      │   ├┄names: ["id", "val"]
      │   ├┄level: 2
      │   └┄name: "first child"
```

Reply:
@wch and I had gone back and forth about this before. 
I think the idea being that attributes fundamentally are just special children of an object rather than a separate data structure so nesting them is slightly misleading. (I could be completely wrong about that though).

Reply:
An earlier implementation had attributes as a subtree, but we decided on the current way for the following reasons:

* It's slightly misleading to show `attributes` as a subtree data structure.
* It reduces visual clutter and it's more vertically compact to show the attributes the way they currently are. For example:

```
x <- factor(c("A", "B"))
tree(x, show_attributes = TRUE)
#> A,B
#> ╰┄attributes
#>   ├─levels:  ["A", "B"]
#>   ╰─class: "factor"
tree(x, show_attributes = TRUE)
#> A,B
#> ├┄<attr>levels: ["A", "B"]
#> ╰┄<attr>class: "factor"
```

It's a bit more verbose on each line, but on the whole it's cleaner the current way, IMO.

Reply:
Ok, that's reasonable. We just need to change the suffix since we use `<>` for types. Maybe `attr(, "levels")` etc?